# coding: utf-8

"""
    

    No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)

    The version of the OpenAPI document: 1.0.0
    Generated by OpenAPI Generator (https://openapi-generator.tech)

    Do not edit the class manually.
"""  # noqa: E501


from __future__ import annotations
import pprint
import re  # noqa: F401
import json


from typing import Any, ClassVar, Dict, List, Optional, Union
from pydantic import BaseModel, StrictFloat, StrictInt, StrictStr
from raassdkpyv2.models.field_errors_value import FieldErrorsValue
try:
    from typing import Self
except ImportError:
    from typing_extensions import Self

class ValidateError(BaseModel):
    """
    ValidateError
    """ # noqa: E501
    name: StrictStr
    message: StrictStr
    stack: Optional[StrictStr] = None
    status: Union[StrictFloat, StrictInt]
    fields: Dict[str, FieldErrorsValue]
    __properties: ClassVar[List[str]] = ["name", "message", "stack", "status", "fields"]

    model_config = {
        "populate_by_name": True,
        "validate_assignment": True,
        "protected_namespaces": (),
    }


    def to_str(self) -> str:
        """Returns the string representation of the model using alias"""
        return pprint.pformat(self.model_dump(by_alias=True))

    def to_json(self) -> str:
        """Returns the JSON representation of the model using alias"""
        # TODO: pydantic v2: use .model_dump_json(by_alias=True, exclude_unset=True) instead
        return json.dumps(self.to_dict())

    @classmethod
    def from_json(cls, json_str: str) -> Self:
        """Create an instance of ValidateError from a JSON string"""
        return cls.from_dict(json.loads(json_str))

    def to_dict(self) -> Dict[str, Any]:
        """Return the dictionary representation of the model using alias.

        This has the following differences from calling pydantic's
        `self.model_dump(by_alias=True)`:

        * `None` is only added to the output dict for nullable fields that
          were set at model initialization. Other fields with value `None`
          are ignored.
        """
        _dict = self.model_dump(
            by_alias=True,
            exclude={
            },
            exclude_none=True,
        )
        # override the default output from pydantic by calling `to_dict()` of each value in fields (dict)
        _field_dict = {}
        if self.fields:
            for _key in self.fields:
                if self.fields[_key]:
                    _field_dict[_key] = self.fields[_key].to_dict()
            _dict['fields'] = _field_dict
        return _dict

    @classmethod
    def from_dict(cls, obj: Dict) -> Self:
        """Create an instance of ValidateError from a dict"""
        if obj is None:
            return None

        if not isinstance(obj, dict):
            return cls.model_validate(obj)

        _obj = cls.model_validate({
            "name": obj.get("name"),
            "message": obj.get("message"),
            "stack": obj.get("stack"),
            "status": obj.get("status"),
            "fields": dict(
                (_k, FieldErrorsValue.from_dict(_v))
                for _k, _v in obj.get("fields").items()
            )
            if obj.get("fields") is not None
            else None
        })
        return _obj


